Tribler/community/doubleentry/community.py

```diff
@@ -0,0 +1,325 @@
+                            distribution=(self.claim_global_time(),),
+                            destination=(signature_request.candidate,),
+                            # Append the signature to the data
+                            payload=data_without_signature + (signature,))
+        return message
+
+    def validate_signature_response(self, signature_response):
+        """
+        Validates if the signature_response message contains correct signatures and public keys.
+        :param signature_response: The message that needs to be checked.
+        :return: Boolean containing the truth value of the validity of the message.
+        """
+        payload = signature_response.payload
+        # Check if the request part is valid.
+        valid_request = self.validate_signature(payload.public_key_requester, payload.signature_data_requester(),
+                                                payload.signature_requester)
+        # Check if the response part is valid.
+        valid_response = self.validate_signature(payload.public_key_responder, payload.signature_data_responder(),
+                                                 payload.signature_responder)
+
+        # Both have to be valid.
+        return valid_request and valid_response
+
+    def _check_signature_response(self, messages):
+        """
+        Generator that generates a list of correct signature_response messages from a list of messages.
+        :param messages: Potential valid signature_response messages.
+        :return: Series of correct signature_response messages
+        """
+        self._logger.info("Received " + str(len(messages)) + " signature responses.")
+        for message in messages:
+            # For now do no checking.
+            if self.validate_signature_response(message):
+                self._logger.info("Received valid response")
+                yield message
+            else:
+                DropMessage(message, "Invalid signature response message")
+
+    def _on_signature_response(self, messages):
+        """
+        Handles behaviour of the community when it receives a signature_response message.
+        It persist the signature response.
+        :param messages: Signature_response messages that needs to be handled.
+        """
+        for message in messages:
+            # Check if the messages are not from ourselves.
+            if message.payload.public_key_responder != self._public_key:
+                self.persist_signature_response(message)
+
+    def persist_signature_response(self, message):
+        """
+        Persist the signature response message.
+        A hash will be created from the message and this will be used as an unique identifier.
+        :param message:
+        """
+        message_hash = self.hash_signature_response(message)
+        self._logger.info("Persisting sr: %s." % base64.encodestring(message_hash))
+        self.persistence.add_block(message_hash, message.payload)
```

Comment:
This could be moved but would result in duplicate code in _on_signature_response and publish_signature_response_message.